File: Tests/MDifyCoreTests/NativeOCRRoutingClientTests.swift
import Foundation
import XCTest
@testable import MDifyCore

final class NativeOCRRoutingClientTests: XCTestCase {
    func testLiteImageUsesNativeVisionAndDoesNotCallWorker() async throws {
        let directory = try makeTemporaryDirectory()
        defer { try? FileManager.default.removeItem(at: directory) }
        let inputURL = directory.appendingPathComponent("scan.png")
        FileManager.default.createFile(atPath: inputURL.path, contents: Data())
        let outputURL = directory.appendingPathComponent("scan.md")
        let worker = SpyWorkerClient(markdown: "# Worker\n")
        let native = StubNativeOCR(result: NativeOCRResult(markdown: "Native recognized text\n", averageConfidence: 0.9))
        let client = NativeOCRRoutingClient(workerKind: .lite, workerClient: worker, nativeOCR: native)

        let response = try await client.convert(inputURL: inputURL, outputURL: outputURL)

        XCTAssertEqual(response.worker, "native")
        XCTAssertEqual(response.engine, "apple-vision")
        XCTAssertTrue(response.ocrUsed)
        XCTAssertEqual(try String(contentsOf: outputURL, encoding: .utf8), "Native recognized text\n")
        let workerInvocationCount = await worker.invocationCount
        XCTAssertEqual(workerInvocationCount, 0)
    }

    func testLiteImageKeepsWeakNonEmptyNativeResultWithWarning() async throws {
        let directory = try makeTemporaryDirectory()
        defer { try? FileManager.default.removeItem(at: directory) }
        let inputURL = directory.appendingPathComponent("scan.png")
        FileManager.default.createFile(atPath: inputURL.path, contents: Data())
        let outputURL = directory.appendingPathComponent("scan.md")
        let worker = SpyWorkerClient(markdown: "# Worker\n")
        let native = StubNativeOCR(result: NativeOCRResult(markdown: "tiny\n", averageConfidence: 0.2))
        let client = NativeOCRRoutingClient(workerKind: .lite, workerClient: worker, nativeOCR: native)

        let response = try await client.convert(inputURL: inputURL, outputURL: outputURL)

        XCTAssertEqual(response.worker, "native")
        XCTAssertTrue(response.warnings.contains { $0.contains("Apple Vision OCR confidence was low") })
        XCTAssertEqual(try String(contentsOf: outputURL, encoding: .utf8), "tiny\n")
        let workerInvocationCount = await worker.invocationCount
        XCTAssertEqual(workerInvocationCount, 0)
    }

    func testLiteImageFailsWhenNativeOCRReturnsNoText() async throws {
        let directory = try makeTemporaryDirectory()
        defer { try? FileManager.default.removeItem(at: directory) }
        let inputURL = directory.appendingPathComponent("blank.png")
        FileManager.default.createFile(atPath: inputURL.path, contents: Data())
        let outputURL = directory.appendingPathComponent("blank.md")
        let worker = SpyWorkerClient(markdown: "# Worker\n")
        let native = StubNativeOCR(result: NativeOCRResult(markdown: "", averageConfidence: 0))
        let client = NativeOCRRoutingClient(workerKind: .lite, workerClient: worker, nativeOCR: native)

        do {
            _ = try await client.convert(inputURL: inputURL, outputURL: outputURL)
            XCTFail("Expected empty native OCR to fail in Lite")
        } catch let error as NativeOCRRoutingError {
            XCTAssertEqual(error, .nativeOCRReturnedNoText)
        }

        let workerInvocationCount = await worker.invocationCount
        XCTAssertEqual(workerInvocationCount, 0)
        XCTAssertFalse(FileManager.default.fileExists(atPath: outputURL.path))
    }

    func testOCRImageFallsBackToRapidOCRWhenNativeResultIsWeak() async throws {
        let directory = try makeTemporaryDirectory()
        defer { try? FileManager.default.removeItem(at: directory) }
        let inputURL = directory.appendingPathComponent("scan.jpg")
        FileManager.default.createFile(atPath: inputURL.path, contents: Data())
        let outputURL = directory.appendingPathComponent("scan.md")
        let preflight = SpyWorkerClient(markdown: "# Preflight\n")
        let fallback = SpyWorkerClient(markdown: "# RapidOCR\n", engine: "rapidocr", ocrUsed: true)
        let native = StubNativeOCR(result: NativeOCRResult(markdown: "tiny\n", averageConfidence: 0.2))
        let client = NativeOCRRoutingClient(
            workerKind: .ocr,
            workerClient: preflight,
            rapidOCRWorkerClient: fallback,
            nativeOCR: native
        )

        let response = try await client.convert(inputURL: inputURL, outputURL: outputURL)

        XCTAssertEqual(response.worker, "ocr")
        XCTAssertEqual(response.engine, "rapidocr")
        XCTAssertTrue(response.warnings.contains { $0.contains("Apple Vision OCR was weak") })
        XCTAssertEqual(try String(contentsOf: outputURL, encoding: .utf8), "# RapidOCR\n")
        let preflightInvocationCount = await preflight.invocationCount
        let fallbackInvocationCount = await fallback.invocationCount
        XCTAssertEqual(preflightInvocationCount, 0)
        XCTAssertEqual(fallbackInvocationCount, 1)
    }

    func testTextPDFUsesMarkItDownPreflightAndSkipsNativeOCR() async throws {
        let directory = try makeTemporaryDirectory()
        defer { try? FileManager.default.removeItem(at: directory) }
        let inputURL = directory.appendingPathComponent("text.pdf")
        FileManager.default.createFile(atPath: inputURL.path, contents: Data())
        let outputURL = directory.appendingPathComponent("text.md")
        let worker = SpyWorkerClient(markdown: "This PDF already has enough extracted text.\n")
        let native = StubNativeOCR(result: NativeOCRResult(markdown: "Native should not run\n", averageConfidence: 0.9))
        let client = NativeOCRRoutingClient(workerKind: .ocr, workerClient: worker, nativeOCR: native)

        let response = try await client.convert(inputURL: inputURL, outputURL: outputURL)

        XCTAssertEqual(response.worker, "ocr")
        XCTAssertEqual(response.engine, "markitdown")
        XCTAssertFalse(response.ocrUsed)
        let workerInvocationCount = await worker.invocationCount
        let nativeInvocationCount = await native.invocationCount
        XCTAssertEqual(workerInvocationCount, 1)
        XCTAssertEqual(nativeInvocationCount, 0)
    }

    func testScannedPDFUsesNativeOCRAndFallsBackToRapidOCRWhenNativeIsWeak() async throws {
        let directory = try makeTemporaryDirectory()
        defer { try? FileManager.default.removeItem(at: directory) }
        let inputURL = directory.appendingPathComponent("scan.pdf")
        FileManager.default.createFile(atPath: inputURL.path, contents: Data())
        let outputURL = directory.appendingPathComponent("scan.md")
        let preflight = SpyWorkerClient(markdown: "\n")
        let fallback = SpyWorkerClient(markdown: "# Rapid PDF\n", engine: "rapidocr", ocrUsed: true)
        let native = StubNativeOCR(result: NativeOCRResult(markdown: "bad\n", averageConfidence: 0.3))
        let client = NativeOCRRoutingClient(
            workerKind: .ocr,
            workerClient: preflight,
            rapidOCRWorkerClient: fallback,
            nativeOCR: native
        )

        let response = try await client.convert(inputURL: inputURL, outputURL: outputURL)

        XCTAssertEqual(response.engine, "rapidocr")
        let preflightInvocationCount = await preflight.invocationCount
        let nativeInvocationCount = await native.invocationCount
        let fallbackInvocationCount = await fallback.invocationCount
        XCTAssertEqual(preflightInvocationCount, 1)
        XCTAssertEqual(nativeInvocationCount, 1)
        XCTAssertEqual(fallbackInvocationCount, 1)
    }

    private func makeTemporaryDirectory() throws -> URL {
        let directory = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory
    }
}

private actor SpyWorkerClient: WorkerConverting {
    private(set) var invocationCount = 0
    private let markdown: String
    private let engine: String
    private let ocrUsed: Bool

    init(markdown: String, engine: String = "markitdown", ocrUsed: Bool = false) {
        self.markdown = markdown
        self.engine = engine
        self.ocrUsed = ocrUsed
    }

    func convert(inputURL: URL, outputURL: URL) async throws -> WorkerResponse {
        invocationCount += 1
        try FileManager.default.createDirectory(at: outputURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        try markdown.write(to: outputURL, atomically: true, encoding: .utf8)
        return WorkerResponse(
            ok: true,
            outputPath: outputURL.path,
            inputPath: inputURL.path,
            worker: ocrUsed ? "ocr" : "ocr",
            engine: engine,
            ocrUsed: ocrUsed,
            warnings: [],
            errorCode: nil,
            message: nil
        )
    }
}

private actor StubNativeOCR: NativeOCRRecognizing {
    private(set) var invocationCount = 0
    private let result: NativeOCRResult

    init(result: NativeOCRResult) {
        self.result = result
    }

    func recognize(inputURL: URL) async throws -> NativeOCRResult {
        invocationCount += 1
        return result
    }
}

File: windows/MDify.Windows.Tests/NativeOcrRoutingClientTests.cs
using MDify.Windows.Models;
using MDify.Windows.Services;

namespace MDify.Windows.Tests;

public sealed class NativeOcrRoutingClientTests : IDisposable
{
    private readonly string _root = Path.Combine(Path.GetTempPath(), $"mdify-routing-{Guid.NewGuid():N}");

    public NativeOcrRoutingClientTests()
    {
        Directory.CreateDirectory(_root);
    }

    [Fact]
    public async Task Image_UsesNativeOcrFirst()
    {
        var input = Touch("scan.png");
        var output = Path.Combine(_root, "scan.md");
        var worker = new SpyWorker("worker text");
        var native = new StubNativeOcr(new NativeOcrResult("Native recognized text", 0.9));
        var client = new NativeOcrRoutingClient(WorkerKind.Lite, worker, nativeOcr: native);

        var response = await client.ConvertAsync(input, output, CancellationToken.None);

        Assert.Equal("native", response.Worker);
        Assert.Equal("windows-text-recognizer", response.Engine);
        Assert.True(response.OcrUsed);
        Assert.Equal("Native recognized text", File.ReadAllText(output));
        Assert.Equal(1, native.InvocationCount);
        Assert.Equal(0, worker.InvocationCount);
    }

    [Fact]
    public async Task Pdf_WithEnoughPreflightText_SkipsNativeOcr()
    {
        var input = Touch("text.pdf");
        var output = Path.Combine(_root, "text.md");
        var worker = new SpyWorker("This PDF already has extractable text.");
        var native = new StubNativeOcr(new NativeOcrResult("Native should not run", 0.9));
        var client = new NativeOcrRoutingClient(WorkerKind.Ocr, worker, nativeOcr: native);

        var response = await client.ConvertAsync(input, output, CancellationToken.None);

        Assert.Equal("ocr", response.Worker);
        Assert.Equal("markitdown", response.Engine);
        Assert.False(response.OcrUsed);
        Assert.Equal(1, worker.InvocationCount);
        Assert.Equal(0, native.InvocationCount);
    }

    [Fact]
    public async Task Pdf_WithAlmostEmptyPreflight_RunsNativeOcr()
    {
        var input = Touch("scan.pdf");
        var output = Path.Combine(_root, "scan.md");
        var worker = new SpyWorker("");
        var native = new StubNativeOcr(new NativeOcrResult("Native PDF text", 0.9));
        var client = new NativeOcrRoutingClient(WorkerKind.Ocr, worker, nativeOcr: native);

        var response = await client.ConvertAsync(input, output, CancellationToken.None);

        Assert.Equal("native", response.Worker);
        Assert.Equal("Native PDF text", File.ReadAllText(output));
        Assert.Equal(1, worker.InvocationCount);
        Assert.Equal(1, native.InvocationCount);
    }

    [Fact]
    public async Task WeakNativeOcr_FallsBackToRapidOcrWorker()
    {
        var input = Touch("scan.jpg");
        var output = Path.Combine(_root, "scan.md");
        var preflight = new SpyWorker("preflight text");
        var fallback = new SpyWorker("# RapidOCR", engine: "rapidocr", ocrUsed: true);
        var native = new StubNativeOcr(new NativeOcrResult("bad", 0.2));
        var client = new NativeOcrRoutingClient(WorkerKind.Ocr, preflight, fallback, native);

        var response = await client.ConvertAsync(input, output, CancellationToken.None);

        Assert.Equal("rapidocr", response.Engine);
        Assert.Contains(response.Warnings, warning => warning.Contains("weak", StringComparison.OrdinalIgnoreCase));
        Assert.Equal("# RapidOCR", File.ReadAllText(output));
        Assert.Equal(0, preflight.InvocationCount);
        Assert.Equal(1, fallback.InvocationCount);
    }

    [Fact]
    public async Task WeakNativeOcr_IsClearErrorWhenRapidOcrFallbackMissing()
    {
        var input = Touch("scan.jpg");
        var output = Path.Combine(_root, "scan.md");
        var preflight = new SpyWorker("preflight text");
        var native = new StubNativeOcr(new NativeOcrResult("bad", 0.2));
        var client = new NativeOcrRoutingClient(WorkerKind.Ocr, preflight, nativeOcr: native);

        var error = await Assert.ThrowsAsync<NativeOcrRoutingException>(() =>
            client.ConvertAsync(input, output, CancellationToken.None));

        Assert.Contains("RapidOCR fallback", error.Message, StringComparison.OrdinalIgnoreCase);
        Assert.Equal(0, preflight.InvocationCount);
    }

    [Fact]
    public async Task NativeFailure_FallsBackToRapidOcrWhenAvailable()
    {
        var input = Touch("scan.png");
        var output = Path.Combine(_root, "scan.md");
        var fallback = new SpyWorker("# RapidOCR", engine: "rapidocr", ocrUsed: true);
        var native = new ThrowingNativeOcr();
        var client = new NativeOcrRoutingClient(WorkerKind.Ocr, new SpyWorker(""), fallback, native);

        var response = await client.ConvertAsync(input, output, CancellationToken.None);

        Assert.Equal("rapidocr", response.Engine);
        Assert.Contains(response.Warnings, warning => warning.Contains("failed", StringComparison.OrdinalIgnoreCase));
    }

    [Fact]
    public async Task NativeFailure_IsClearErrorWithoutRapidOcrFallback()
    {
        var input = Touch("scan.png");
        var output = Path.Combine(_root, "scan.md");
        var client = new NativeOcrRoutingClient(
            WorkerKind.Lite,
            new SpyWorker(""),
            nativeOcr: new ThrowingNativeOcr());

        var error = await Assert.ThrowsAsync<NativeOcrRoutingException>(() =>
            client.ConvertAsync(input, output, CancellationToken.None));

        Assert.Contains("Windows Text Recognizer", error.Message, StringComparison.OrdinalIgnoreCase);
    }

    public void Dispose()
    {
        if (Directory.Exists(_root))
        {
            Directory.Delete(_root, recursive: true);
        }
    }

    private string Touch(string fileName)
    {
        var path = Path.Combine(_root, fileName);
        File.WriteAllText(path, "");
        return path;
    }

    private sealed class StubNativeOcr : INativeOcrService
    {
        private readonly NativeOcrResult _result;

        public StubNativeOcr(NativeOcrResult result)
        {
            _result = result;
        }

        public int InvocationCount { get; private set; }

        public Task<NativeOcrResult> RecognizeAsync(string inputPath, CancellationToken cancellationToken)
        {
            InvocationCount++;
            return Task.FromResult(_result);
        }
    }

    private sealed class ThrowingNativeOcr : INativeOcrService
    {
        public Task<NativeOcrResult> RecognizeAsync(string inputPath, CancellationToken cancellationToken)
        {
            throw new InvalidOperationException("model unavailable");
        }
    }

    private sealed class SpyWorker : IWorkerConverting
    {
        private readonly string _markdown;
        private readonly string _engine;
        private readonly bool _ocrUsed;

        public SpyWorker(string markdown, string engine = "markitdown", bool ocrUsed = false)
        {
            _markdown = markdown;
            _engine = engine;
            _ocrUsed = ocrUsed;
        }

        public WorkerKind WorkerKind => WorkerKind.Ocr;

        public int InvocationCount { get; private set; }

        public Task<WorkerResponse> ConvertAsync(
            string inputPath,
            string outputPath,
            CancellationToken cancellationToken)
        {
            InvocationCount++;
            Directory.CreateDirectory(Path.GetDirectoryName(outputPath)!);
            File.WriteAllText(outputPath, _markdown);
            return Task.FromResult(new WorkerResponse(
                true,
                outputPath,
                inputPath,
                "ocr",
                _engine,
                _ocrUsed,
                Array.Empty<string>(),
                null,
                null));
        }
    }
}
